File: backend/csharp/.net/microservices/ms-cnpj-validacao/CnpjValidator.cs
using System.Text.RegularExpressions;

namespace MsCnpjValidacao;

public static class CnpjValidator
{
    private static readonly Regex NonDigits = new(@"\D", RegexOptions.Compiled);

    public static bool IsValid(string? input)
    {
        var normalized = Normalize(input);
        if (normalized is null)
            return false;

        if (normalized.Length != 14)
            return false;

        if (AllDigitsSame(normalized))
            return false;

        var d1 = CalculateCheckDigit(normalized, 12);
        var d2 = CalculateCheckDigit(normalized, 13);

        return normalized[12] == (char)('0' + d1)
            && normalized[13] == (char)('0' + d2);
    }

    public static string? Normalize(string? input)
    {
        if (input is null)
            return null;

        var digitsOnly = NonDigits.Replace(input, "");
        if (string.IsNullOrWhiteSpace(digitsOnly))
            return null;

        return digitsOnly;
    }

    private static bool AllDigitsSame(string digits)
    {
        var first = digits[0];
        for (var i = 1; i < digits.Length; i++)
        {
            if (digits[i] != first)
                return false;
        }
        return true;
    }

    private static int CalculateCheckDigit(string digits, int length)
    {
        int[] weights = length == 12
            ? new[] { 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 }
            : new[] { 6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2 };

        var sum = 0;
        for (var i = 0; i < length; i++)
        {
            var digit = digits[i] - '0';
            sum += digit * weights[i];
        }

        var mod = sum % 11;
        return mod < 2 ? 0 : 11 - mod;
    }
}

File: backend/java/spring/microservices/ms-cnpj-validacao/src/main/java/br/com/estudo/cnpj/CnpjValidator.java
package br.com.estudo.cnpj;

import java.util.Objects;

public final class CnpjValidator {

    private CnpjValidator() {
    }

    public static boolean isValid(String input) {
        String normalized = normalize(input);
        if (normalized == null) {
            return false;
        }

        if (normalized.length() != 14) {
            return false;
        }

        if (allDigitsSame(normalized)) {
            return false;
        }

        int d1 = calculateCheckDigit(normalized, 12);
        int d2 = calculateCheckDigit(normalized, 13);

        return normalized.charAt(12) == (char) ('0' + d1)
                && normalized.charAt(13) == (char) ('0' + d2);
    }

    public static String normalize(String input) {
        if (input == null) {
            return null;
        }
        String digitsOnly = input.replaceAll("\\D", "");
        if (digitsOnly.isBlank()) {
            return null;
        }
        return digitsOnly;
    }

    private static boolean allDigitsSame(String digits) {
        char first = digits.charAt(0);
        for (int i = 1; i < digits.length(); i++) {
            if (digits.charAt(i) != first) {
                return false;
            }
        }
        return true;
    }

    private static int calculateCheckDigit(String digits, int length) {
        Objects.checkFromIndexSize(0, length, digits.length());

        int[] weights = (length == 12)
                ? new int[]{5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2}
                : new int[]{6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};

        int sum = 0;
        for (int i = 0; i < length; i++) {
            int digit = digits.charAt(i) - '0';
            sum += digit * weights[i];
        }

        int mod = sum % 11;
        return (mod < 2) ? 0 : (11 - mod);
    }
}
